renderlib/FileReaderZarr.cpp

```diff
@@ -0,0 +1,340 @@
+#include <map>
+#include <set>
+
+FileReaderZarr::FileReaderZarr() {}
+
+FileReaderZarr::~FileReaderZarr() {}
+
+static ::nlohmann::json
+jsonRead(std::string zarrurl)
+{
+  // JSON uses a separate driver
+  auto attrs_store =
+    tensorstore::Open<::nlohmann::json, 0>(
+      { { "driver", "json" }, { "kvstore", { { "driver", "http" }, { "base_url", zarrurl }, { "path", ".zattrs" } } } })
+      .result()
+      .value();
+
+  // Sets attrs_array to a rank-0 array of ::nlohmann::json
+  auto attrs_array_result = tensorstore::Read(attrs_store).result();
+
+  ::nlohmann::json attrs;
+  if (attrs_array_result.ok()) {
+    attrs = attrs_array_result.value()();
+    std::cout << "attrs: " << attrs << std::endl;
+  } else if (absl::IsNotFound(attrs_array_result.status())) {
+    attrs = ::nlohmann::json::object_t();
+  } else {
+    std::cout << "Error: " << attrs_array_result.status();
+  }
+  return attrs;
+}
+
+uint32_t
+FileReaderZarr::loadNumScenesZarr(const std::string& filepath)
+{
+  return 1;
+}
+
+// return number of bytes copied to dest
+static size_t
+copyDirect(uint8_t* dest, const uint8_t* src, size_t numBytes, int srcBitsPerPixel)
+{
+  memcpy(dest, src, numBytes);
+  return numBytes;
+}
+
+// convert pixels
+// this assumes tight packing of pixels in both buf(source) and dataptr(dest)
+// assumes dest is of format IN_MEMORY_BPP
+// return 1 for successful conversion, 0 on failure (e.g. unacceptable srcBitsPerPixel)
+static size_t
+convertChannelData(uint8_t* dest, const uint8_t* src, const VolumeDimensions& dims)
+{
+  // how many pixels in this channel:
+  size_t numPixels = dims.sizeX * dims.sizeY * dims.sizeZ;
+  int srcBitsPerPixel = dims.bitsPerPixel;
+
+  // dest bits per pixel is IN_MEMORY_BPP which is currently 16, or 2 bytes
+  if (ImageXYZC::IN_MEMORY_BPP == srcBitsPerPixel) {
+    memcpy(dest, src, numPixels * (srcBitsPerPixel / 8));
+    return 1;
+  } else if (srcBitsPerPixel == 8) {
+    uint16_t* dataptr16 = reinterpret_cast<uint16_t*>(dest);
+    for (size_t b = 0; b < numPixels; ++b) {
+      *dataptr16 = (uint16_t)src[b];
+      dataptr16++;
+    }
+    return 1;
+  } else if (srcBitsPerPixel == 32) {
+    // assumes 32-bit floating point (not int or uint)
+    uint16_t* dataptr16 = reinterpret_cast<uint16_t*>(dest);
+    const float* src32 = reinterpret_cast<const float*>(src);
+    // compute min and max; and then rescale values to fill dynamic range.
+    float lowest = FLT_MAX;
+    float highest = -FLT_MAX;
+    float f;
+    for (size_t b = 0; b < numPixels; ++b) {
+      f = src32[b];
+      if (f < lowest) {
+        lowest = f;
+      }
+      if (f > highest) {
+        highest = f;
+      }
+    }
+    for (size_t b = 0; b < numPixels; ++b) {
+      *dataptr16 = (uint16_t)((src32[b] - lowest) / (highest - lowest) * 65535.0);
+      dataptr16++;
+    }
+    return 1;
+  } else {
+    LOG_ERROR << "Unexpected tiff pixel size " << srcBitsPerPixel << " bits";
+    return 0;
+  }
+  return 0;
```

Comment:
Unreachable `return` - could either remove or break the code out of the `else` block above